### ISSUE TITLE
Logs flooded with "dropped sample" errors

### ISSUE THREAD

**Issue by slapcat** [canonical/grafana-agent-k8s-operator#296]
### Bug Description

System logs are flooded with the following error leading to high disk utilization:

```
caller=dedupe.go:112 agent=prometheus level=info url=http://10.242.34.71/cos-prometheus-0/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling"
```

There should be a way to trim these down or filter them.

### To Reproduce

1. Deploy grafana-agent charm.
2. Relate to compatible machine charm.
3. Watch /var/log/syslog for error.

### Environment

Grafana-agent is related to machine charms in an openstack cloud.

### Relevant log output

```shell
caller=dedupe.go:112 agent=prometheus level=info url=http://10.242.34.71/cos-prometheus-0/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling"
```

**Comment by PietroPasotti** [comment]
*edit find these logs in the grafana-agent logs

**Comment by IbraAoad** [comment]
Might be related https://github.com/prometheus/prometheus/issues/11589

**Comment by ca-scribner** [comment]
The linked issue form @IbraAoad suggests the problem is backpressure (good summary comment [here](https://github.com/prometheus/prometheus/issues/11589#issuecomment-1625100368)).  I tried to reproduce this but was unable to get "Sample dropped" errors.  

First I tried just using a simple machine charm (`ubuntu`) but that didn't cause any issues for the grafana-agent machine charm.  

Second I tried grafana-agent-k8s, using flog and avalanche to send large amounts of fake logs/metrics.  I artificially created backpressure by continually killed prometheus and loki (`while true; do kubectl delete pod prometheus-0; kubectl delete pod loki-0; sleep 1; done`), but still couldn't reproduce the "Sample dropped" errors.  

Do you have a minimal reproduction environment I could use?  

Also, within your environment, it would be useful to have:
* the grafana-agent config file
* these metrics when the problem is occurring: 
  * `curl -s <grafana-agent-unit>:12345/metrics | grep "prometheus_remote_storage_enqueue_retries_total"`
  * `curl -s <grafana-agent-unit>:12345/metrics | grep "prometheus_remote_storage_shards_max"`

Which revision of the charm are you using?  A related fix upstream was posted in agent 0.35.2, but that version is only in our agent charm

**Comment by sed-i** [comment]
If you could also include the output of the following at normal operation and at time of failure, that could be interesting.

```bash
juju ssh grafana-agent/0 curl -s localhost:12345/metrics | grep -E "wal|agent_inflight_requests|log_messages_total|failed" | grep -v "^# "

juju ssh gragna-agent/0 df -h
```

**Comment by ca-scribner** [comment]
I might have been wrong about where this patch landed.  [the fix commit](https://github.com/grafana/agent/commit/3837d47dbbfff039b49c5a8272a97c52feb4b58b) is tagged >=v0.36.0, but looks like it was also cherry-picked into [v0.35.2](https://github.com/grafana/agent/pull/4624/commits) and shows up in the [v0.35.2 changelog](https://github.com/grafana/agent/blob/v0.36.0/CHANGELOG.md#v0352-2023-07-27).  So probably it is fixed in v0.35.2.  

Either way, the next steps are:
1. @slapcat: if you can, provide a good reproduction case
2. observability: update grafana-agent to something more recent (~0.40.x)

I'm working on (2) now

**Comment by ca-scribner** [comment]
To address this, we've done the following:
* bump the grafana-agent workload to 0.40.4 for [k8s](https://github.com/canonical/grafana-agent-k8s-operator/pull/303) and [machine](https://github.com/canonical/grafana-agent-operator/pull/142) charms.  Now all charms and architectures are all on the same 0.40.4 version
* some other updates, like [pinning the grafana-agent workload in machine charms](https://github.com/canonical/grafana-agent-operator/pull/131) so the grafana-agent workload version is predictable

Based on those fixes, we're closing this issue as we think it should be addressed.  We were unsuccessful in actually reproducing the issue, however, so please reopen if you face this again!
